src/gtest/test_proofs.cpp

```diff
@@ -0,0 +1,553 @@
+        auto proof = libsnark::r1cs_ppzksnark_prover<curve_pp>(
+            kp.pk,
+            example.primary_input,
+            example.auxiliary_input,
+            example.constraint_system
+        );
+
+        ASSERT_TRUE(libsnark::r1cs_ppzksnark_verifier_strong_IC<curve_pp>(
+            kp.vk,
+            example.primary_input,
+            proof
+        ));
+
+        ZCProof compressed_proof_0(proof);
+
+        CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+        ss << compressed_proof_0;
+
+        ZCProof compressed_proof_1;
+        ss >> compressed_proof_1;
+
+        ASSERT_TRUE(compressed_proof_0 == compressed_proof_1);
+
+        auto newproof = compressed_proof_1.to_libsnark_proof<libsnark::r1cs_ppzksnark_proof<curve_pp>>();
+
+        ASSERT_TRUE(proof == newproof);
+        ASSERT_TRUE(libsnark::r1cs_ppzksnark_verifier_strong_IC<curve_pp>(
+            kp.vk,
+            example.primary_input,
+            newproof
+        ));
+    }
+}
+
+TEST(proofs, g1_deserialization)
+{
+    CompressedG1 g;
+    curve_G1 expected;
+
+    // Valid G1 element.
+    {
+        CDataStream ss(ParseHex("0230644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd46"), SER_NETWORK, PROTOCOL_VERSION);
+        ss >> g;
+
+        expected.X = curve_Fq("21888242871839275222246405745257275088696311157297823662689037894645226208582");
+        expected.Y = curve_Fq("3969792565221544645472939191694882283483352126195956956354061729942568608776");
+        expected.Z = curve_Fq::one();
+
+        ASSERT_TRUE(g.to_libsnark_g1<curve_G1>() == expected);
+    }
+
+    // Its negation.
+    {
+        CDataStream ss(ParseHex("0330644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd46"), SER_NETWORK, PROTOCOL_VERSION);
+        ss >> g;
+
+        expected.X = curve_Fq("21888242871839275222246405745257275088696311157297823662689037894645226208582");
+        expected.Y = curve_Fq("3969792565221544645472939191694882283483352126195956956354061729942568608776");
+        expected.Z = curve_Fq::one();
+
+        ASSERT_TRUE(g.to_libsnark_g1<curve_G1>() == -expected);
+    }
+
+    // Invalid leading bytes
+    {
+        CDataStream ss(ParseHex("ff30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd46"), SER_NETWORK, PROTOCOL_VERSION);
+
+        ASSERT_THROW(ss >> g, std::ios_base::failure);
+    }
+
+    // Invalid point
+    {
+        CDataStream ss(ParseHex("0208c6d2adffacbc8438f09f321874ea66e2fcc29f8dcfec2caefa21ec8c96a77c"), SER_NETWORK, PROTOCOL_VERSION);
+        ss >> g;
+
+        ASSERT_THROW(g.to_libsnark_g1<curve_G1>(), std::runtime_error);
+    }
+
+    // Point with out of bounds Fq
+    {
+        CDataStream ss(ParseHex("02ffc6d2adffacbc8438f09f321874ea66e2fcc29f8dcfec2caefa21ec8c96a77c"), SER_NETWORK, PROTOCOL_VERSION);
+        ss >> g;
+
+        ASSERT_THROW(g.to_libsnark_g1<curve_G1>(), std::logic_error);
+    }
+
+    // Randomly produce valid G1 representations and fail/succeed to
```

Comment:
Why only valid representations? This test would work even for random byte strings, correct? You'd want to make the first byte valid to ensure that a significant proportion of strings are exercising the decompression code, but I think q is close enough to 2<sup>256</sup> that the rest can be random. Also, the compressed y bit should be random.